libdebug/snapshots/diff.py

```diff
@@ -0,0 +1,283 @@
+#
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from libdebug.snapshots.memory.memory_map_diff import MemoryMapDiff
+from libdebug.snapshots.memory.memory_map_diff_list import MemoryMapDiffList
+from libdebug.snapshots.registers.register_diff import RegisterDiff
+from libdebug.snapshots.registers.register_diff_accessor import RegisterDiffAccessor
+from libdebug.utils.platform_utils import get_platform_register_size
+from libdebug.utils.pprint_primitives import (
+    pprint_diff_line,
+    pprint_diff_substring,
+    pprint_inline_diff,
+    pprint_memory_diff_util,
+    pprint_reg_diff_large_util,
+    pprint_reg_diff_util,
+)
+
+if TYPE_CHECKING:
+    from libdebug.snapshots.registers.snapshot_registers import SnapshotRegisters
+    from libdebug.snapshots.snapshot import Snapshot
+
+
+class Diff:
+    """This object represents a diff between two snapshots."""
+
+    def __init__(self: Diff, snapshot1: Snapshot, snapshot2: Snapshot) -> None:
+        """Initialize the Diff object with two snapshots.
+
+        Args:
+            snapshot1 (Snapshot): The first snapshot.
+            snapshot2 (Snapshot): The second snapshot.
+        """
+        if snapshot1.snapshot_id < snapshot2.snapshot_id:
+            self.snapshot1 = snapshot1
+            self.snapshot2 = snapshot2
+        else:
+            self.snapshot1 = snapshot2
+            self.snapshot2 = snapshot1
+
+        # The level of the diff is the lowest level among the two snapshots
+        if snapshot1.level == "base" or snapshot2.level == "base":
+            self.level = "base"
+        elif snapshot1.level == "writable" or snapshot2.level == "writable":
+            self.level = "writable"
+        else:
+            self.level = "full"
+
+        if self.snapshot1.arch != self.snapshot2.arch:
+            raise ValueError("Snapshots have different architectures. Automatic diff is not supported.")
+
+    def _save_reg_diffs(self: Snapshot) -> None:
+        self.regs = RegisterDiffAccessor(
+            self.snapshot1.regs._generic_regs,
+            self.snapshot1.regs._special_regs,
+            self.snapshot1.regs._vec_fp_regs,
+        )
+
+        all_regs = dir(self.snapshot1.regs)
+        all_regs = [reg for reg in all_regs if isinstance(self.snapshot1.regs.__getattribute__(reg), int | float)]
+
+        for reg_name in all_regs:
+            old_value = self.snapshot1.regs.__getattribute__(reg_name)
+            new_value = self.snapshot2.regs.__getattribute__(reg_name)
+            has_changed = old_value != new_value
+
+            diff = RegisterDiff(
+                old_value=old_value,
+                new_value=new_value,
+                has_changed=has_changed,
+            )
+
+            # Create diff object
+            self.regs.__setattr__(reg_name, diff)
+
+    def _resolve_maps_diff(self: Diff) -> None:
+        # Handle memory maps
+        all_maps_diffs = []
+        handled_map2_indices = []
```

Comment:
This could maybe go from a list to a set? Not sure if it does anything for performance in Python